File: api/src/test/java/net/blueshell/api/factory/model/ContributionFactory.java
package net.blueshell.api.factory.model;

import com.github.javafaker.Faker;
import lombok.RequiredArgsConstructor;
import net.blueshell.api.model.User;
import net.blueshell.api.model.contribution.Contribution;
import net.blueshell.api.model.contribution.ContributionPeriod;
import org.springframework.stereotype.Component;

import java.util.concurrent.atomic.AtomicLong;

/**
 * Factory for Contribution model test instances.
 */
@Component
@RequiredArgsConstructor
public class ContributionFactory {

    private static final AtomicLong counter = new AtomicLong(1000);
    private final Faker faker;
    private final UserFactory userFactory;
    private final ContributionPeriodFactory contributionPeriodFactory;

    public Contribution createBasic() {
        Contribution c = new Contribution();
        c.setId(generateId());
        User user = userFactory.createBasic();
        ContributionPeriod period = contributionPeriodFactory.createBasic();
        c.setUser(user);
        c.setUserId(user.getId());
        c.setContributionPeriod(period);
        c.setContributionPeriodId(period.getId());
        return c;
    }

    public Contribution createFull() {
        return createBasic();
    }

    public Contribution createWithCustomizations(java.util.function.Consumer<Contribution> customizer) {
        Contribution c = createFull();
        customizer.accept(c);
        return c;
    }

    private Long generateId() {
        return counter.incrementAndGet();
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/model/contribution/ContributionFactory.kt
package net.blueshell.api.factory.model.contribution

import com.github.javafaker.Faker
import net.blueshell.api.domain.contribution.persistence.Contribution
import net.blueshell.api.factory.model.UserFactory
import org.springframework.stereotype.Component
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Factory for Contribution model test instances.
 */
@Component
class ContributionFactory(
    private val faker: Faker,
    private val userFactory: UserFactory,
    private val contributionPeriodFactory: ContributionPeriodFactory
) {

    fun createBasic(): net.blueshell.api.domain.contribution.persistence.Contribution {
        val contribution = _root_ide_package_.net.blueshell.api.domain.contribution.persistence.Contribution()
        val user = userFactory.createBasic()
        val period = contributionPeriodFactory.createBasic()
        contribution.user = user
        contribution.contributionPeriod = period
        return contribution
    }

    fun createFull(): net.blueshell.api.domain.contribution.persistence.Contribution = createBasic()

    fun createWithCustomizations(customizer: Consumer<net.blueshell.api.domain.contribution.persistence.Contribution>): net.blueshell.api.domain.contribution.persistence.Contribution {
        val contribution = createFull()
        customizer.accept(contribution)
        return contribution
    }

    private fun generateId(): Long = COUNTER.incrementAndGet()

    private companion object {
        val COUNTER = AtomicLong(1000)
    }
}
